packages/frontend/workspace/src/providers/index.ts

```diff
@@ -1,160 +1,129 @@
-  const promise = new Promise<void>((resolve, reject) => {
-    _resolve = resolve;
-    _reject = reject;
-  });
+export * from './sync';
 
-  return {
-    flavour: 'local-indexeddb',
-    active: true,
-    get whenReady() {
-      return promise;
-    },
-    cleanup: () => {
-      // todo: cleanup data
-    },
-    sync: () => {
-      logger.info('sync indexeddb provider', id);
-      datasource
-        .queryDocState(doc.guid, {
-          stateVector: encodeStateVector(doc),
-        })
-        .then(docState => {
-          if (docState) {
-            Y.applyUpdate(doc, docState.missing, indexedDBDownloadOrigin);
+export const createLocalProviders = (): DocProviderCreator[] => {
+  return [
+    (_, doc, { awareness }) => {
+      const engine = new SyncEngine(
+        doc,
+        environment.isDesktop
+          ? createSQLiteStorage(doc.guid)
+          : createIndexedDBStorage(doc.guid),
+        []
+      );
+
+      const awarenessProviders = [
+        createBroadcastChannelAwarenessProvider(doc.guid, awareness),
+      ];
+
+      let connected = false;
+
+      return {
+        flavour: '_',
+        passive: true,
+        active: true,
+        sync() {
+          if (!connected) {
+            engine.start();
+
+            for (const provider of awarenessProviders) {
+              provider.connect();
+            }
+            connected = true;
           }
-          _resolve();
-        })
-        .catch(_reject);
+        },
+        get whenReady() {
+          return engine.waitForLoadedRootDoc();
+        },
+        connect() {
+          // TODO: actually connect
+        },
+        disconnect() {
+          // TODO: actually disconnect
+        },
+        get connected() {
+          return connected;
+        },
+        engine,
+      };
     },
-  };
-};
-
-export {
-  createAffineSocketIOProvider,
-  createBroadcastChannelProvider,
-  createIndexedDBBackgroundProvider,
-  createIndexedDBDownloadProvider,
-  createSQLiteDBDownloadProvider,
-  createSQLiteProvider,
-  downloadBinaryFromCloud,
-};
-
-export const createLocalProviders = (): DocProviderCreator[] => {
-  const providers = [
-    createIndexedDBBackgroundProvider,
-    createIndexedDBDownloadProvider,
-  ] as DocProviderCreator[];
-
-  if (runtimeConfig.enableBroadcastChannelProvider) {
-    providers.push(createBroadcastChannelProvider);
-  }
-
-  if (environment.isDesktop && runtimeConfig.enableSQLiteProvider) {
-    providers.push(createSQLiteProvider, createSQLiteDBDownloadProvider);
-  }
-
-  return providers;
+  ];
 };
 
 export const createAffineProviders = (): DocProviderCreator[] => {
-  return (
-    [
-      ...createLocalProviders(),
-      runtimeConfig.enableCloud && createAffineSocketIOProvider,
-      runtimeConfig.enableCloud && createMergeCloudSnapshotProvider,
-    ] as DocProviderCreator[]
-  ).filter(v => Boolean(v));
+  return [
+    (_, doc, { awareness }) => {
+      const engine = new SyncEngine(
```

Comment:
I meet issues of opening a cloud workspace without data persisted

- workspace title becomes `Untitled` in first load, going well in later loads
- `Upgrade workspace data` shown in first load, going well in later loads

These are exactly the issues I mentioned by `switching local` feature in design

Reply:
The second has been fixed.

I plan to solve the first issue when I refactor the workspace manager